lib/Dialect/DC/Transforms/DCMaterialization.cpp

```diff
@@ -0,0 +1,142 @@
+//===- DCMaterialization.cpp - Fork/sink materialization pass ---*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Contains the definitions of the Fork/sink materialization pass.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Dialect/DC/DCOps.h"
+#include "circt/Dialect/DC/DCPasses.h"
+
+using namespace circt;
+using namespace dc;
+using namespace mlir;
+
+static void replaceFirstUse(Operation *op, Value oldVal, Value newVal) {
+  for (int i = 0, e = op->getNumOperands(); i < e; ++i)
+    if (op->getOperand(i) == oldVal) {
+      op->setOperand(i, newVal);
+      break;
+    }
+  return;
+}
+
+static void insertSink(Value val, OpBuilder &rewriter) {
+  rewriter.setInsertionPointAfterValue(val);
+  rewriter.create<SinkOp>(val.getLoc(), val);
+}
+
+static void insertFork(Value result, OpBuilder &rewriter) {
+  // Get successor operations
+  std::vector<Operation *> opsToProcess;
+  for (auto &u : result.getUses())
+    opsToProcess.push_back(u.getOwner());
+
+  // Insert fork after op
+  rewriter.setInsertionPointAfterValue(result);
+  auto forkSize = opsToProcess.size();
+  auto newFork = rewriter.create<ForkOp>(result.getLoc(), result, forkSize);
+
+  // Modify operands of successor
+  // opsToProcess may have multiple instances of same operand
+  // Replace uses one by one to assign different fork outputs to them
+  for (auto [op, forkRes] : llvm::zip(opsToProcess, newFork->getResults()))
+    replaceFirstUse(op, result, forkRes);
+}
+
+// Insert Fork Operation for every operation and function argument with more
+// than one successor.
+static LogicalResult addForkOps(Region &r, OpBuilder &rewriter) {
+  for (Operation &op : r.getOps()) {
+    // Ignore terminators, and don't add Forks to Forks.
+    if (op.getNumSuccessors() == 0 && !isa<ForkOp>(op)) {
```

Comment:
Isn't it possible that a `ForkOp` instance that was present in the input IR before the fork/sink materialization pass runs has one of its results used multiple times? In that case, it seems we could end up with a value used more than once after the pass is finished. Does the pass assume that no forks/sinks are present in the input IR?

Reply:
It shouldn't assume that IMO.

Reply:
Most definitely - to create a vague excuse for myself, this code is just duplicated from the Handshake stuff. I'll fix it and add a test for it, good catch 👍

Reply:
This actually exposed an issue in the fork-to-fork canonicalizer - will fix in a later comit.